### ISSUE TITLE
Comparison with upcoming tensorflow.contrib.mpi_collectives

### ISSUE THREAD

**Issue by chenliu0831** [uber/horovod#23]
Just curious if you guys notice a recent contribution from [here](https://github.com/tensorflow/tensorflow/pull/12299) which added `allreduce/allgather` (which I think horovod derived from?) and a similar optimizer wrapper. I'm planning to do a benchmark with both horovod and above.  It'll be great if you guys could share some insight in terms of features/flexibility/performance.

Thanks for the great library!

**Comment by alsrgv** [comment]
Hi, @chenliu0831.

This is a great question. Indeed, Horovod was based on allreduce code from Baidu. At the time when Horovod was started, Baidu's fork appeared stale, so we decided to take it upon ourselves to improve it.

There are few changes that we made:

1. We converted the code to be stand-alone Python package. We wanted to make it extremely easy and quick for folks to try out Horovod, so we didn't want them to have to convert to the latest version of TensorFlow, install Bazel and rebuild TensorFlow from source. Instead, Horovod takes seconds to install.

2. We replaced Baidu *ring-allreduce* implementation with NVIDIA NCCL. NCCL allowed us to enjoy significant performance boost thanks to all the optimizations that NVIDIA put into it. To use NCCL, a user just needs to pass a couple of [environment variables](https://github.com/uber/horovod#have-gpus) to pip install.

3. We added support for models that fit inside a single server, potentially on multiple GPUs, whereas original version only supported models that fit on single GPU.

4. We made a number of API improvements that were motivated by a feedback we received from a number of initial users. A couple of examples:

   1. We implemented a broadcast operation that enforces consistent initialization of the model on all the workers.

   2. We made `hvd.size()`, `hvd.rank()`, `hvd.local_rank()` Python native ops, rather than TensorFlow ops, which makes it easy to pin GPU to be used:

      ```python
      config.gpu_options.visible_device_list = str(hvd.local_rank())
      ```

      Original Baidu's counterparts are TensorFlow ops, which introduces chicken-egg problem - TensorFlow op would create a session that will allocate memory on all GPUs, but you need to run that op to determine which GPU to pin your session to.

5. We further improved performance with [Tensor Fusion](https://github.com/uber/horovod#tensor-fusion). Depending on the model and hardware, we got up to 20% performance improvement out of it.

6. We added [Horovod Timelines](https://github.com/uber/horovod#analyzing-horovod-performance) to analyze performance bottlenecks. Tensor Fusion was a result of looking at a timeline of ResNet-152 run and realizing how many tiny tensors it's trying to independently reduce.

Please share your findings from your performance comparison.

**Comment by chenliu0831** [comment]
@alsrgv Thanks! Those are very helpful.

**Comment by chenliu0831** [comment]
@alsrgv Hey I might have a stupid question. At [this line] (https://github.com/alsrgv/benchmarks/commit/86bf2f9269dbefb4e57a8b66ed260c8fab84d6c7#diff-632d3c06e6607601ccb136629596a464R920) of your fork of tensorflow benchmarks, it seems to make it only 1 shard, which seems confusing to me how the data parallelism is achieved.  Is there anything I'm missing or I misunderstand the input pipeline?

**Comment by alsrgv** [comment]
@chenliu0831 the simple solution is to use regular TensorFlow data pipeline, like https://www.tensorflow.org/api_guides/python/reading_data or https://www.tensorflow.org/performance/performance_models#input_pipeline. These pipelines randomize the order in which files are read and add random shuffling to TFRecords, which means that on average each TFRecord will be read the same number of times in random order.

The benchmark uses [`data_flow_ops.RecordInput`](https://www.tensorflow.org/performance/performance_models#parallelize_io_reads) from latter guide, e.g.: https://github.com/alsrgv/benchmarks/blob/horovod/scripts/tf_cnn_benchmarks/preprocessing.py#L360. The benchmark does fix `seed=301` for reproducibility, so if you do want to use it for model training you'd have to remove `seed` attribute as well as `tf.set_random_seed(1234)` and `np.random.seed(4321)` to truly randomize it.

**Comment by chenliu0831** [comment]
@alsrgv gotcha Thanks!

**Comment by alsrgv** [comment]
@chenliu0831, any luck with your tests?

**Comment by chenliu0831** [comment]
@alsrgv Hey Alsrgv, sry I didn't get enough bandwidth on this yet (partially because I need to get approval on open source MPI license check).

**Comment by alsrgv** [comment]
Gotcha, good luck with getting the approval!

**Comment by chenliu0831** [comment]
Ah I doubt I'll have bandwidth to actually prioritize this in a short term. If someone else already done it or interested in this, feel free to take over or close this issue. Sorry.

**Comment by alsrgv** [comment]
No worries. Feel free to reopen this issue or raise a new issue when you have more questions.
